### PR TITLE
fix: fix ios safari edge and cancelable issue 

### PR DESCRIPTION
**Pull request by daybrush**
## Issue

#146 
#147

### CODE REVIEW

File: src/inputType/PanInput.ts [hunk 1]
```diff
@@ -236,20 +239,61 @@ export class PanInput implements IInputType {
 	protected onHammerInput(event) {
 		if (this.isEnable()) {
 			if (event.isFirst) {
-				this.observer.hold(this, event);
+				this.panFlag = false;
+
+				if (event.srcEvent.cancelable !== false) {
+					this.observer.hold(this, event);
+					this.isRightEdge = IS_IOS_SAFARI && event.center.x > window.innerWidth - 20;
```
Comment (mixed) [inline src/inputType/PanInput.ts:246]: I think it would be good if `20` is constant.

File: src/inputType/PanInput.ts [hunk 2]
```diff
@@ -236,20 +239,61 @@ export class PanInput implements IInputType {
...
+		}
 		const userDirection = getDirectionByAngle(
 			event.angle, this.options.thresholdAngle);
 
 		// not support offset properties in Hammerjs - start
 		const prevInput = this.hammer.session.prevInput;
+		const clientX = event.center.x;
+
+		if (prevInput && IS_IOS_SAFARI) {
+			if (clientX < 0) {
```
Comment (mixed) [inline src/inputType/PanInput.ts:267]: Why don't you delete the comment and replace it with a variable?
```ts
const SwipeLeftToRight = clientX < 0;
if (SwipeLeftToRight) {
}
```

File: src/inputType/PanInput.ts [hunk 3]
```diff
@@ -236,20 +239,61 @@ export class PanInput implements IInputType {
...
+					velocityY: 0,
+					offsetX: 0,
+					offsetY: 0,
+				});
+				return;
+			} else if (this.isRightEdge) {
+				clearTimeout(this.rightEdgeTimer);
 
+				// - is right to left
+				if (event.deltaX < -20) {
```
Comment (mixed) [inline src/inputType/PanInput.ts:281]: Why don't you delete the comment and replace it with a variable?
```ts
const SwipeRightToLeft = event.deltaX < -20;
if (SwipeRightToLeft) {
}
```

File: src/inputType/PanInput.ts [hunk 4]
```diff
@@ -265,15 +309,25 @@ export class PanInput implements IInputType {
 				useDirection(DIRECTION_VERTICAL, this._direction, userDirection),
 			]);
 		const prevent = offset.some(v => v !== 0);
+
 		if (prevent) {
-			event.srcEvent.preventDefault();
-			event.srcEvent.stopPropagation();
+			const srcEvent = event.srcEvent;
+
+			if (srcEvent.cancelable !== false) {
```
Comment (mixed) [inline src/inputType/PanInput.ts:316]: I think that if statement unnecessary because preventDefault works only when `srcEvent.cancelable` is true.

Comment (daybrush) [reply src/inputType/PanInput.ts:316]: @mixed 

Yes. But there are browsers without cancelable.

Comment (mixed) [reply src/inputType/PanInput.ts:319]: @daybrush 
OK. If so, `srcEvent.cancelable === true` seems to be more readable.

Comment (daybrush) [reply src/inputType/PanInput.ts:319]: @mixed 

If there is no cancelable, preventDefault is possible. It is possible except when it is false.

File: src/inputType/PanInput.ts [hunk 5]
```diff
@@ -236,20 +239,64 @@ export class PanInput implements IInputType {
 	protected onHammerInput(event) {
 		if (this.isEnable()) {
 			if (event.isFirst) {
-				this.observer.hold(this, event);
+				this.panFlag = false;
+
+				if (event.srcEvent.cancelable !== false) {
+					this.observer.hold(this, event);
+					this.isRightEdge = IS_IOS_SAFARI && event.center.x > window.innerWidth - EDGE_DISTANCE;
+					this.panFlag = true;
```
Comment (WoodNeck) [inline src/inputType/PanInput.ts:247]: The name `panFlag` looks not meaningful. How about naming it to its purpose? Like, `scrolling` or so?

File: src/inputType/PanInput.ts [hunk 6]
```diff
@@ -236,20 +239,64 @@ export class PanInput implements IInputType {
...
+			return;
+		}
 		const userDirection = getDirectionByAngle(
 			event.angle, this.options.thresholdAngle);
 
 		// not support offset properties in Hammerjs - start
 		const prevInput = this.hammer.session.prevInput;
 
+		if (prevInput && IS_IOS_SAFARI) {
+			const SwipeLeftToRight = event.center.x < 0;
```
Comment (WoodNeck) [inline src/inputType/PanInput.ts:266]: Why do these swipe-related variables use PascalCase, unlike the other ones?

File: src/inputType/PanInput.ts (hunk 5 above)
Comment (daybrush) [reply src/inputType/PanInput.ts:247]: @WoodNeck 

A `panmove` occurs even after a forced release from the axes, but a `change` event occurs by accepting it in observer.change.

**Pull request merged**